pkg/components/pubsub/pluggable.go

```diff
@@ -0,0 +1,210 @@
+		m := pubsub.NewMessage{
+			Data:        msg.Data,
+			ContentType: &msg.ContentType,
+			Topic:       msg.Topic,
+			Metadata:    msg.Metadata,
+		}
+		var ackError *proto.AckMessageError
+
+		if err := handler(ctx, &m); err != nil {
+			p.logger.Errorf("error when handling message on topic %s", msg.Topic)
+			ackError = &proto.AckMessageError{
+				Message: err.Error(),
+			}
+		}
+
+		// As per documentation:
+		// When using streams,
+		// one must take care to avoid calling either SendMsg or RecvMsg multiple times against the same Stream from different goroutines.
+		// In other words, it's safe to have a goroutine calling SendMsg and another goroutine calling RecvMsg on the same stream at the same time.
+		// But it is not safe to call SendMsg on the same stream in different goroutines, or to call RecvMsg on the same stream in different goroutines.
+		// https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md#streams
+		sendLock.Lock()
+		if err := streamingPull.Send(&proto.PullMessagesRequest{
+			AckMessageId: msg.Id,
+			AckError:     ackError,
+		}); err != nil {
+			p.logger.Errorf("error when ack'ing message %s from topic %s", msg.Id, msg.Topic)
+		}
+		sendLock.Unlock()
+	}
+}
+
+// pullMessages pull messages of the given subscription and execute the handler for that messages.
+func (p *grpcPubSub) pullMessages(ctx context.Context, subscription *proto.Subscription, handler pubsub.Handler) error {
+	// first pull should be sync and subsequent connections can be made in background if necessary
+	pull, err := p.Client.PullMessages(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "unable to subscribe")
+	}
+
+	streamCtx, cancel := context.WithCancel(pull.Context())
+
+	err = pull.Send(&proto.PullMessagesRequest{
+		Subscription: subscription,
+	})
+
+	cleanup := func() {
+		if closeErr := pull.CloseSend(); closeErr != nil {
+			p.logger.Warnf("could not close pull stream of topic %s: %v", subscription.Topic, closeErr)
+		}
+		cancel()
+	}
+
+	if err != nil {
+		cleanup()
+		return errors.Wrapf(err, "unable to subscribe")
+	}
+
+	handle := p.adaptHandler(streamCtx, pull, handler)
+	go func() {
+		defer cleanup()
+		for {
+			msg, err := pull.Recv()
+			if err == io.EOF { // no more messages
+				return
+			}
+
+			// TODO reconnect on error
+			if err != nil {
+				p.logger.Errorf("failed to receive message: %v", err)
+				return
+			}
+
+			p.logger.Debugf("received message from stream on topic %s", msg.Topic)
+
+			go handle(msg)
+		}
+	}()
+
+	return nil
+}
+
+// Subscribe subscribes to a given topic and callback the handler when a new message arrives.
+func (p *grpcPubSub) Subscribe(ctx context.Context, req pubsub.SubscribeRequest, handler pubsub.Handler) error {
+	subscription := &proto.Subscription{
+		Topic:    req.Topic,
+		Metadata: req.Metadata,
+	}
+	return p.pullMessages(ctx, subscription, handler)
+}
+
+func (p *grpcPubSub) Close() error {
+	return p.GRPCConnector.Close()
```

Comment:
Qq is there any chance of the connector being nil?

Reply:
If there is, will fail for the other functions as well, notice this is an embedded field which can be accessed directly from p.X()

I'm going to remove this since it's redundant by the way, thanks!